### ISSUE TITLE
CLI version suggestion: --version-semver

### ISSUE THREAD

**Issue by AloisMahdal** [mojombo/semver#362]
Problem
=====

It's a well-known convention for CLI tools to include `--version` option that only causes the program to report its version and exit.  It's not so common, however, to have the version information in predictable format.  For example:

    $ python --version
    Python 2.7.13
    $ tr --version
    tr (GNU coreutils) 8.25
    Copyright (C) 2016 Free Software Foundation, Inc.
    License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
    This is free software: you are free to change and redistribute it.
    There is NO WARRANTY, to the extent permitted by law.

    Written by Jim Meyering.
    $ curl --version
    curl 7.51.0 (x86_64-redhat-linux-gnu) libcurl/7.51.0 NSS/3.27 zlib/1.2.8 libidn2/0.16 libpsl/0.17.0 (+libidn2/0.11) libssh2/1.8.0 nghttp2/1.13.0
    Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp 
    Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz HTTP2 UnixSockets Metalink PSL 
    $ 
Having extra info in the text may be useful but unless common format is kept, it's hard to consume the version number by scripts such as test harnesses or management tools.

Convention
=======

Now some authors may decide to be more cooperative by implementing SemVer.  These will obviously show SemVer-compatible version in the text.  A nice convention would be to provide also option `--version-semver`, which would be bound to ouput *only the version* and nothing else.  This option would normally accompany the "traditional" `--version` option, which could still include extra "free form" detail.

**Comment by zafarkhaja** [comment]
--**ver**sion-sem**ver**, isn't it a duplication, how about --semver?

**Comment by krzysiekpiasecki** [comment]
>> --semver?

semver of tool?

**Comment by AloisMahdal** [comment]
It's intended to be subset of the `--version` namespace.  Version info is a very specific meta-datum, so it's natural to have specific way of getting it.  Also it's not really intended for humans (eg. interactive mode) so there's no need to keep it short.

Calling it `--semver` is problematic.  `semver` is IMO way too short to be recommendable for "all CLI tools in general".  It might nor sound probable but this particular cluster of six letters could have already other, completely unrelated meaning: "version" is not the only word starting with "ver" and "sem" is not the only word starting with "sem", and that's just one way of splitting it.

A special case of this could be a tool which actually works in the domain of version information and management.  Such tool could naturally already use this switch (or want to keep the option to use it for later).

This gets even worse when you realize that SemVer is not the only versioning standard out there.  This convention could be pretty easily extended to any number of other versioning standards. But changing the convention to use the name *directly* breaks this extensibility.

(Also, you are a bit mixing apples with oranges here. It's no more duplication than phrase "SemVer version":  both "version" and "semver" have different meaning; they rather *specify* each other rather than just repeat.)

**Comment by FichteFoll** [comment]
Just make it - - semantic-version imo. I wanted to comment on the overall proposal as well, but lack the time currently.

**Comment by AloisMahdal** [comment]
@FichteFoll Thanks for the comment but it does not seem convincing.  You don't seem to reveal much about reasoning behind why "--semantic-version" would be better than my proposal "--version-semver". As I said in last comment, it's intentional to make it *subset* of the existing well-known namespace.

**Comment by krzysiekpiasecki** [comment]
@AloisMahdal The convention you are proposing is a bad idea. You propose an additional interface to return version information. Simplicity is achieved by removing interfaces rather than adding more in the name of greater flexibility. In addition, the presented script assumes that we take action according to the full version. This does not include components that will still need to be parsed anyway.

**Comment by AloisMahdal** [comment]
@krzysiekpiasecki I agree that the less is better. This is actually meant for people who *want more* but rather than everyone come up with own spelling (`--semver`, `--semantic-version`), the "less" is in offering universal spelling that could work for all. So in its way this is application of the same principle.

About the "want more": the main motivation for this was in context of automated testing. I was developing a test harness and I wanted to show version of the tested module as part of the log, but keep the parsing code as trivial as possible (as every test-related code should strive for) without sacrificing the conventional `--version` which I already used for human-readable information. It's worth noting that there was already assumption that the version itself does say enough; if that's not the case, you will indeed designate yourself.

(That said, your comment did make me think a bit more critical about this proposal.)

**Comment by jwdonahue** [comment]
This is an interesting idea that belongs to the realm of things that layer on top of SemVer.  Until such layers are established and adopted, there is no point in attempting to be proscriptive from within the SemVer spec.  Let's let SemVer be one simple node in the web of evolving dev-ops standards, keep it simple, but not too simple.  

I vote to close this issue.

**Comment by AloisMahdal** [comment]
@jwdonahue See Note in the header of the original post. By any stretch, this was never intended to be part of SemVer.  Original SemVer points to this issue tracker as a means to "leave feedback" and as you can see, the tracker has been already used for many years as medium for discussions about both SemVer and conventions built above it.

**Comment by jwdonahue** [comment]
@AloisMahdal, what/where is this "Original SemVer"?  Please forgive my ignorance, I am still stumbling around a bit here at GitHub.  When I see open issues, I see technical debt and have a strong desire to drive it to zero. I am all for drawing attention here to other sites and ideas, I just don't see the point in leaving this on the list of SemVer open issues. Until I chimed in, there has been no activity on this issue since April.

**Comment by AloisMahdal** [comment]
@jwdonahue welcome to GitHub. Good spirit is welcome, but please note that this is a very diverse communty with very diverse practices; basically each project can have its own rules. So don't fall into the illusion that the only reason for an issue being open (and inactive) is because people forgot about it. It's perfectly OK for an issue to be open for years; not all issues represent a technical debt.

Generally, people will be happy if you want to contribute to their project but they will want you to get familiar with contributing guidelines first. (If an issue is affecting you, though, it's OK to just politely ask about the state :-)).

(Note: I'm not even representing community around SemVer; It was only after carefully reading many other issues and discussion that I decided that the tracker might be the right place for my own suggestion.)

In short, we are all guests here.

**Comment by jwdonahue** [comment]
@AloisMahdal, I've [started another issue to discuss this](https://github.com/semver/semver/issues/413), rather than further pollute your topic.

**Comment by AloisMahdal** [comment]
FTR, I have to admit that over past years I wrote dozens of CLI's but I haven't implemented this convention anywhere.  I always used `-V` for -- just (parseable) SemVer, and `--version` for perhaps something more chatty.

I think the best argument against my past self would be: if `--version-semver` only works if consumer knows it exists (and what it means).  Ie. there has to be pre-existing contract about CLI.  If one wants to have such contract then it could be `--version` or `-V` right away, coupled with expectation that it has to be SemVer data.

The use case I had in mind when proposing this could still arguably make use of it:  I was imagining a framework that would wrap CLI's in other interfaces, so for reporting version there would have to be such interface.  (And `--version` could be already "occupied" by a "chatty version", and `-V` is too short.)

I think the reason I've never used this is that for all practical purposes it works way better to communicate version info using proper config management or package management, such as rpm, dpkg, etc.  If only because it works regardless of CLI.  IOW, it's unlikely that such framework would exist without relying on proper management like that so using CLI for that would be duplication.

(OTOH, that's also why I never used `module.VERSION` convention in Python, and yet I think their packaging still supports it.)
